Containerfile

```diff
@@ -3,57 +3,70 @@
 
 
-# Stores source cache
-FROM shuttle-build as cache
-ARG PROD
+# Stores source cache and cargo chef recipe
+FROM cargo-chef as planner
 WORKDIR /src
 COPY . .
-RUN find ${SRC_CRATES} \( -name "*.proto" -or -name "*.rs" -or -name "*.toml" -or -name "Cargo.lock" -or -name "README.md" -or -name "*.sql" -or -name "ulid0.so" \) -type f -exec install -D \{\} /build/\{\} \;
-# This is used to carry over in the docker images any *.pem files from shuttle root directory,
-# to be used for TLS testing, as described here in the admin README.md.
-RUN if [ "$PROD" != "true" ]; then \
-    find ${SRC_CRATES} -name "*.pem" -type f -exec install -D \{\} /build/\{\} \;; \
-    fi
-
-
-# Stores cargo chef recipe
-FROM shuttle-build AS planner
-COPY --from=cache /build .
-RUN cargo chef prepare --recipe-path recipe.json
-
+# Select only the essential files for copying into next steps
+# so that changes to miscellaneous files don't trigger a new cargo-chef cook.
+# Beware that .dockerignore filters files before they get here.
+RUN find . \( \
+    -name "*.rs" -or \
+    -name "*.toml" -or \
+    -name "Cargo.lock" -or \
+    -name "*.sql" -or \
+    # Used for local TLS testing, as described in admin/README.md
+    -name "*.pem" -or \
+    -name "ulid0.so" \
+    \) -type f -exec install -D \{\} /build/\{\} \;
+WORKDIR /build
+RUN cargo chef prepare --recipe-path /recipe.json
+# TODO upstream: Reduce the cooking by allowing multiple --bin args to prepare, or like this https://github.com/LukeMathWalker/cargo-chef/issues/181
 
-# Builds crate according to cargo chef recipe
-FROM shuttle-build AS builder
-ARG folder
-COPY --from=planner /build/recipe.json recipe.json
-RUN cargo chef cook --release --recipe-path recipe.json
 
-COPY --from=cache /build .
-RUN cargo build --bin shuttle-${folder} --release
+# Builds crate according to cargo chef recipe.
+# This step is skipped if the recipe is unchanged from previous build (no dependencies changed).
+FROM cargo-chef AS builder
+ARG CARGO_PROFILE
+COPY --from=planner /recipe.json /
+# https://i.imgflip.com/2/74bvex.jpg
+RUN cargo chef cook \
+    --all-features \
+    $(if [ "$CARGO_PROFILE" = "release" ]; then echo --release; fi) \
+    --recipe-path /recipe.json
+COPY --from=planner /build .
+# Building all at once to share build artifacts in the "cook" layer
+RUN cargo build \
+    $(if [ "$CARGO_PROFILE" = "release" ]; then echo --release; fi) \
+    --bin shuttle-auth \
+    --bin shuttle-deployer \
+    --bin shuttle-provisioner \
+    --bin shuttle-gateway \
+    --bin shuttle-resource-recorder \
+    --bin shuttle-next -F next
 
 
-# The final image for this "shuttle-..." crate
+# The final image for running each "shuttle-..." binary
 ARG RUSTUP_TOOLCHAIN
 FROM docker.io/library/rust:${RUSTUP_TOOLCHAIN}-buster as shuttle-crate
+ARG CARGO_PROFILE
 ARG folder
+ARG crate
 ARG prepare_args
 # used as env variable in prepare script
 ARG PROD
-ARG RUSTUP_TOOLCHAIN
-ENV RUSTUP_TOOLCHAIN=${RUSTUP_TOOLCHAIN}
```

Comment:
Did you move this or remove it? This is actually a fix for a strange bug we used to have, where users projects with dependencies that had toolchain overrides cause the build to fail with a version mismatch. It's been a while though, and I'm not sure we still have this issue now that we don't use the cargo library. We used to be able to reliably reproduce it by deploying a serenity project with the cache feature enabled (which pulled in dashmap).